expression/builtin_string.go

```diff
@@ -3976,3 +3979,111 @@ func (b *builtinWeightStringSig) evalString(row chunk.Row) (string, bool, error)
 	}
 	return string(ctor.Key(str)), false, nil
 }
+
+type soundexFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *soundexFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, types.ETString)
+	if err != nil {
+		return nil, err
+	}
+	argTp := args[0].GetType()
+	bf.tp.Flen = argTp.Flen
+	SetBinFlagOrBinStr(argTp, bf.tp)
+	sig := &builtinSoundexSig{bf}
+	sig.setPbCode(tipb.ScalarFuncSig_Soundex)
+	return sig, nil
+}
+
+type builtinSoundexSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinSoundexSig) Clone() builtinFunc {
+	newSig := &builtinSoundexSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalString evals SOUNDEX(str)
+// See: https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_soundex
+func (b *builtinSoundexSig) evalString(row chunk.Row) (string, bool, error) {
+	str, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if err != nil {
+		return "", false, err
+	}
+	if isNull {
+		return "", true, nil
+	}
+	return soundex(str), false, nil
+}
+
+var soundexMappings = map[string][]string{
```

Comment:
Why not define the soundexTable directly?
``` go
soundexTable = map[string]string{
"B": "1",
"F": "1',
...
}
```

Reply:
IMO, The form `soundexMappings` is more readable and its correctness can easily be checked through https://en.wikipedia.org/wiki/Soundex. What do you think?